ta/gatekeeper/gatekeeper_ta.c

```diff
@@ -0,0 +1,767 @@
+
+	return TA_ComputeSignature(signature, signature_length,
+				   key, salted_password,
+				   sizeof(salted_password));
+}
+
+static TEE_Result TA_CreatePasswordHandle(struct password_handle
+					  *pass_handle,
+					  salt_t salt, secure_id_t user_id,
+					  uint64_t flags,
+					  uint64_t handle_version,
+					  const uint8_t *password,
+					  uint32_t password_length)
+{
+	struct password_handle pw_handle;
+	const uint32_t metadata_length = sizeof(pw_handle.user_id) +
+					 sizeof(pw_handle.flags) +
+					 sizeof(pw_handle.version);
+	uint8_t to_sign[password_length + metadata_length];
+
+	TEE_ObjectHandle master_key = TEE_HANDLE_NULL;
+	TEE_Result res;
+
+	res = TEE_AllocateTransientObject(TEE_TYPE_HMAC_SHA256,
+					  TEE_SHA256_HASH_SIZE * 8,
+					  &master_key);
+	if (res) {
+		EMSG("Failed to allocate password key");
+		goto exit;
+	}
+
+	pw_handle.version = handle_version;
+	pw_handle.salt = salt;
+	pw_handle.user_id = user_id;
+	pw_handle.flags = flags;
+	pw_handle.hardware_backed = true;
+
+	memcpy(to_sign, &pw_handle, metadata_length);
+	memcpy(to_sign + metadata_length, password, password_length);
+
+	res = TA_GetMasterKey(master_key);
+	if (res) {
+		EMSG("Failed to get master key");
+		goto free_key;
+	}
+
+	res = TA_ComputePasswordSignature(pw_handle.signature,
+					  sizeof(pw_handle.signature),
+					  master_key, to_sign,
+					  sizeof(to_sign), salt);
+	if (res) {
+		EMSG("Failed to compute password signature");
+		goto free_key;
+	}
+
+	memcpy(pass_handle, &pw_handle, sizeof(pw_handle));
+
+free_key:
+	TEE_FreeTransientObject(master_key);
+exit:
+	return res;
+}
+
+static TEE_Result TA_GetAuthTokenKey(TEE_ObjectHandle key)
+{
+	TEE_Result res;
+	TEE_Param params[TEE_NUM_PARAMS];
+	TEE_TASessionHandle sess;
+	TEE_Attribute attr;
+	const TEE_UUID uuid = TA_KEYMASTER_UUID;
+
+	uint32_t param_types;
+	uint32_t return_origin;
+
+	uint8_t	dummy[TEE_SHA256_HASH_SIZE];
+	uint8_t	auth_token_key_data[TEE_SHA256_HASH_SIZE];
+
+
+	DMSG("Connect to keymaster");
+
+	param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE);
+	memset(params, 0, sizeof(params));
+
+	res = TEE_OpenTASession(&uuid, TEE_TIMEOUT_INFINITE,
+				param_types, params, &sess,
+				&return_origin);
+	if (res) {
+		EMSG("Failed to connect to keymaster");
+		goto exit;
+	}
+
+	param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				      TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				      TEE_PARAM_TYPE_NONE,
+				      TEE_PARAM_TYPE_NONE);
+	memset(&params, 0, sizeof(params));
+
+	params[0].memref.buffer = dummy;
+	params[0].memref.size = sizeof(dummy);
+
+	params[1].memref.buffer = auth_token_key_data;
+	params[1].memref.size = sizeof(auth_token_key_data);
+
+	res = TEE_InvokeTACommand(sess, TEE_TIMEOUT_INFINITE,
+				  KM_GET_AUTHTOKEN_KEY,
+				  param_types, params, &return_origin);
+	if (res) {
+		EMSG("Failed in keymaster");
```

Comment:
`EMSG("Failed to get authentication token key from Keymaster");`?

Reply:
fixed